### PR TITLE
[WIP] Define getting-started example notebook using TensorFlow & Add support of graph-mode

### LINKED ISSUE #283: [BUG] Add tf graph-mode support
### Steps/Code to reproduce bug

1.  Define the model for NextItemPrediction
2. Compile model with graph model: `model.compile(optimizer="adam", run_eagerly=False)
More details can be found  in this unit [test](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/7aca855103c2243f32f69c7a6452ea67320406b8/tests/tf/model/test_model.py#L107) 

### Expected behavior

Making sure the tf pipeline (training, evaluating, and saving the model) is working in `graph mode`

### PR DESCRIPTION
**Pull request by sararb**
Fixes #283 #262 #258 #236

### Goals :soccer: and Implementation Details :construction:

TL;DR: This PR was initially created to fix the saving of TF model needed by Triton serving  and the definition of Tensorflow version of getting started notebook.  The issue had a larger scope than the saving method and I had to refactor the custom classes `MaskSequence`, `NextItemPredictionTask`, `RankingMetric` and `TransformerBlock`. The main changes are: 

- [x] Update HuggingFace dependency to 4.11* to support Albert and ELECTRA. (Note that ELECTRA model still need feature #263 for working correctly)

- [x] Re-factor the tf ops used to define the custom classes to make sure they are working in both modes: `eager` and `graph`. 

- [x] Define `get_config` / `from_config` methods to the custom classes for model saving and loading. 

- [x] Add a non trainable tf.Variable to `RankingMetric` classes as placeholder to store the updated states at each iteration. This is needed in `graph` mode as the metrics are part of the Keras class `NextItemPredictionTask`.

- [x] Improve the signature HF transformers `call` to fix the issue of shape mismatch when saving the model. This error is linked to non-used arguments being initialized with wrong tensors: The previous call method provided transformer_kwargs twice, e.g. `mask_perm` was set equal to `input_embed` instead of None).  

- [x] Extract from the HF model the serializable layer TF*MainLayer. In fact, all the HF models are defined with one custom Keras layer containing the architecture graph. (e.g. [tfXLNetModel](https://github.com/huggingface/transformers/blob/9e4ea2517504f144e74d9a356d58e9f2be32b3fa/src/transformers/models/xlnet/modeling_tf_xlnet.py#L1160))

- [x] Fix the loss function used by `NextItemPredictionTask`. We set `logits=true` in SparseCategoricalCrossentropy as the task already returns the logit scores. 

- [x] Expose compute_metrics flag of `compute_loss` methods to compute metrics during train / evaluation.  The method `process_metric` is added to flatten the list of scores returned by ranking metrics with respect to their threshold. 

- [x] Create a TensorFlow version of the getting started notebook, including model definition, training, evaluation, saving, and reloading.  

- [ ] Re-load the model for:  model serving, calling the method on a batch of data (`model()`), resume training (`model.fit()`), and compute evaluation scores on new data (`model.evaluate()`)  
  -  Using `.save()` and `tf.keras.models.load_model()` allow for saving the model's architecture and its weights. Applying the re-loaded model is working. However calling `fit()` and `evaluate()` methods raise an input signature error ```ValueError: Could not find matching function to call loaded from the SavedModel.```
  -  The second solution is to use weight saving and loading (`save_weights` / `load_weights`) but we need to construct the model before loading weight.  In this case, we can resume training without any error. 

### Testing Details :mag:

- Add tests to `test_model.py`: 
   - `test_simple_seq_classification`: Add a test of the sequence classification model and check if loss decreasing.
   - `test_next_item_fit`: Test, next item prediction task, using eager and graph modes.
   - `test_simple_model`: Test saving and loading of a simple classification model.
   - `test_save_load_item_prediction`: Test saving and loading of next item prediction task with a simple model using an `MLP` block.
   -  `test_save_load_transformer_model`: Test saving and loading of next item prediction task with supported HF transformers.
   - `test_resume_training`: Test resume training of next item prediction task using weight saving and loading

- Update `test_transformer.py` to include Albert config. 

- Add `test_score_different_from_zero` to `test_ranking_metrics.py` to check that computed metrics are different from 0.  

### Remaining tasks out of scope of this PR: 
- Fix scores of ranking metrics returned by the fit/evaluate methods in the getting-started notebook: The current scores are inconsistent for `mlm` (close to one).  In fact, the model is always returning the last non-padded item suggesting there is a leakage in the model definition (might be related to the masking schema computed by the `TabularSequenceFeatures` module). 

- For `clm`, reported RecallAt seems correct, while NDCGAt is returning the same score for all thresholds (the issue might be related to the nested usage of DCGAt metric)

### CODE REVIEW

**Comment by rnyak** [comment]
@sararb a qq: why do we have the batch as sparse representation instead of dense even though we set `sparse_as_dense=True` in the KerasSequenceLoader. When I print ` next(iter(get_dataloader(train_paths)))[0]['category-list_trim']` out it gives a tuple of (values, nnzs) as below:

```
(<tf.Tensor: shape=(673, 1), dtype=int64, numpy=
 array([[22],
        [ 6],
        [ 3],
        [12],
        [ 3],
        [ 4],
        [ 9],
        [ 9],
        [ 2],
        [ 2],
        [ 4],
        [ 4],
        [ 2],
        [ 8],
        [ 6],
        [ 3],
        [ 4],
        [ 2],
        [ 3],
        [ 9],
        [ 6],
        [25],
        [ 4],
        [ 4],
        [16],
        [ 8],
        [ 2],
        [ 8],
        [18],
        [ 2],
        [ 3],
        [ 8],
        [ 2],
        [ 5],
        [ 4],
        [ 6],
        [ 3],
        [ 3],
        [ 3],
        [ 7],
        [ 6],
        [19],
        [ 5],
        [ 2],
        [11],
        [ 2],
        [ 2],
        [13],
        [ 2],
        [19],
        [ 4],
        [ 4],
        [15],
        [13],
        [28],
        [ 2],
        [11],
        [36],
        [ 7],
        [ 5],
        [ 7],
        [36],
        [ 6],
        [ 4],
        [ 2],
        [12],
        [ 3],
        [ 8],
        [ 4],
        [13],
        [14],
        [18],
        [ 6],
        [ 8],
        [12],
        [ 3],
        [ 4],
        [ 3],
        [11],
        [ 8],
        [19],
        [ 5],
        [ 2],
        [ 4],
        [15],
        [28],
        [ 8],
        [ 5],
        [ 7],
        [ 4],
        [ 2],
        [ 5],
        [ 2],
        [ 5],
        [ 3],
        [ 2],
        [12],
        [11],
        [81],
        [ 2],
        [ 2],
        [ 2],
        [12],
        [ 6],
        [ 6],
        [ 2],
        [ 7],
        [10],
        [ 3],
        [ 5],
        [25],
        [ 4],
        [12],
        [ 3],
        [ 3],
        [ 6],
        [ 2],
        [21],
        [67],
        [ 9],
        [ 6],
        [ 3],
        [ 4],
        [ 3],
        [ 2],
        [13],
        [ 5],
        [58],
        [33],
        [ 2],
        [ 3],
        [ 6],
        [ 5],
        [ 2],
        [ 3],
        [16],
        [ 6],
        [ 2],
        [ 3],
        [ 5],
        [ 2],
        [ 3],
        [ 2],
        [ 2],
        [12],
        [12],
        [ 9],
        [ 7],
        [ 2],
        [ 7],
        [ 4],
        [43],
        [ 5],
        [ 8],
        [ 6],
        [ 8],
        [ 8],
        [ 2],
        [ 4],
        [ 4],
        [ 6],
        [10],
        [ 3],
        [11],
        [19],
        [ 6],
        [ 6],
        [10],
        [ 3],
        [ 6],
        [ 2],
        [ 3],
        [ 5],
        [12],
        [ 9],
        [11],
        [20],
        [ 4],
        [ 2],
        [ 6],
        [ 6],
        [ 9],
        [15],
        [12],
        [10],
        [31],
        [16],
        [10],
        [ 5],
        [38],
        [ 3],
        [ 8],
        [ 5],
        [ 7],
        [ 5],
        [ 2],
        [12],
        [11],
        [ 4],
        [ 5],
        [ 5],
        [ 4],
        [12],
        [ 7],
        [ 7],
        [ 4],
        [ 2],
        [10],
        [ 2],
        [ 5],
        [10],
        [ 7],
        [ 7],
        [11],
        [15],
        [ 4],
        [ 8],
        [16],
        [ 7],
        [ 4],
        [ 3],
        [ 6],
        [ 6],
        [17],
        [ 9],
        [ 5],
        [ 5],
        [16],
        [ 3],
        [31],
        [ 3],
        [29],
        [28],
        [ 4],
        [ 6],
        [ 7],
        [ 3],
        [ 9],
        [ 2],
        [10],
        [ 6],
        [ 9],
        [37],
        [17],
        [ 6],
        [ 2],
        [ 5],
        [ 7],
        [ 4],
        [14],
        [ 2],
        [ 9],
        [ 3],
        [14],
        [29],
        [ 9],
        [ 4],
        [ 3],
        [ 4],
        [ 4],
        [ 7],
        [ 3],
        [ 3],
        [ 4],
        [ 6],
        [ 9],
        [18],
        [ 4],
        [20],
        [ 2],
        [ 7],
        [ 6],
        [ 6],
        [ 2],
        [ 5],
        [ 4],
        [ 7],
        [ 7],
        [ 2],
        [10],
        [ 8],
        [ 5],
        [ 4],
        [ 2],
        [26],
        [13],
        [17],
        [ 5],
        [ 4],
        [ 2],
        [ 2],
        [10],
        [22],
        [ 6],
        [ 6],
        [23],
        [ 2],
        [18],
        [ 4],
        [15],
        [ 9],
        [ 5],
        [ 5],
        [ 5],
        [36],
        [ 3],
        [ 5],
        [43],
        [21],
        [12],
        [ 3],
        [ 5],
        [ 6],
        [ 6],
        [ 2],
        [ 6],
        [ 4],
        [ 4],
        [ 3],
        [17],
        [ 2],
        [11],
        [13],
        [ 4],
        [30],
        [ 7],
        [11],
        [ 4],
        [27],
        [64],
        [ 3],
        [ 4],
        [34],
        [ 7],
        [ 6],
        [ 7],
        [ 3],
        [ 7],
        [ 2],
        [ 9],
        [ 3],
        [ 2],
        [18],
        [10],
        [ 2],
        [ 3],
        [ 4],
        [13],
        [15],
        [ 2],
        [ 2],
        [20],
        [ 4],
        [ 3],
        [ 2],
        [11],
        [10],
        [ 5],
        [10],
        [ 4],
        [ 5],
        [ 4],
        [ 3],
        [ 2],
        [ 2],
        [23],
        [ 3],
        [ 8],
        [ 3],
        [ 2],
        [ 2],
        [ 4],
        [ 6],
        [ 3],
        [ 5],
        [22],
        [ 5],
        [ 3],
        [ 8],
        [ 2],
        [ 5],
        [10],
        [ 4],
        [16],
        [ 5],
        [ 2],
        [ 4],
        [ 3],
        [ 2],
        [ 2],
        [ 3],
        [ 8],
        [13],
        [18],
        [10],
        [ 3],
        [ 3],
        [ 6],
        [ 8],
        [ 3],
        [ 4],
        [ 2],
        [ 6],
        [ 5],
        [12],
        [19],
        [ 8],
        [ 6],
        [20],
        [14],
        [12],
        [ 6],
        [ 3],
        [ 2],
        [ 9],
        [ 9],
        [ 2],
        [10],
        [ 5],
        [ 2],
        [30],
        [ 5],
        [47],
        [ 8],
        [12],
        [10],
        [ 8],
        [ 3],
        [ 4],
        [ 7],
        [ 2],
        [ 4],
        [ 4],
        [ 8],
        [10],
        [21],
        [28],
        [ 4],
        [14],
        [10],
        [ 2],
        [17],
        [ 4],
        [10],
        [ 4],
        [ 9],
        [ 4],
        [ 6],
        [ 2],
        [ 4],
        [ 4],
        [ 2],
        [ 2],
        [ 2],
        [43],
        [14],
        [ 4],
        [ 3],
        [ 3],
        [ 3],
        [ 8],
        [ 4],
        [ 2],
        [18],
        [24],
        [ 2],
        [ 5],
        [ 3],
        [ 3],
        [ 9],
        [19],
        [ 7],
        [ 5],
        [ 9],
        [ 4],
        [ 3],
        [50],
        [10],
        [ 2],
        [ 7],
        [ 2],
        [ 3],
        [ 2],
        [ 4],
        [ 2],
        [ 9],
        [13],
        [ 4],
        [ 5],
        [12],
        [ 3],
        [ 8],
        [ 6],
        [ 4],
        [43],
        [ 4],
        [ 6],
        [ 3],
        [ 3],
        [ 4],
        [ 5],
        [15],
        [ 2],
        [15],
        [12],
        [ 2],
        [ 4],
        [ 2],
        [ 4],
        [ 2],
        [ 8],
        [ 3],
        [ 2],
        [ 3],
        [ 6],
        [ 8],
        [ 4],
        [ 2],
        [10],
        [ 2],
        [ 3],
        [ 5],
        [14],
        [44],
        [ 5],
        [ 2],
        [ 3],
        [ 4],
        [37],
        [ 3],
        [ 4],
        [ 2],
        [ 2],
        [ 3],
        [ 2],
        [ 3],
        [18],
        [ 5],
        [18],
        [ 5],
        [21],
        [15],
        [12],
        [30],
        [ 2],
        [ 3],
        [ 2],
        [ 9],
        [10],
        [ 4],
        [ 3],
        [ 4],
        [ 4],
        [10],
        [ 4],
        [20],
        [ 3],
        [ 5],
        [ 5],
        [ 6],
        [ 7],
        [ 8],
        [24],
        [ 6],
        [ 5],
        [ 3],
        [ 3],
        [12],
        [ 2],
        [ 8],
        [12],
        [15],
        [18],
        [11],
        [ 7],
        [30],
        [ 4],
        [ 4],
        [ 6],
        [15],
        [23],
        [11],
        [ 2],
        [ 4],
        [ 8],
        [ 5],
        [27],
        [ 6],
        [ 6],
        [18],
        [ 3],
        [ 8],
        [ 3],
        [ 7],
        [19],
        [ 2],
        [ 2],
        [ 3],
        [20],
        [ 7],
        [17],
        [17],
        [ 3],
        [ 2],
        [ 5],
        [ 2],
        [ 3],
        [14],
        [ 9],
        [ 4],
        [ 6],
        [ 6],
        [ 5],
        [ 5],
        [ 7],
        [ 4],
        [ 5],
        [ 4],
        [ 4],
        [10],
        [ 3],
        [10],
        [ 2],
        [ 2],
        [ 4],
        [ 5],
        [ 5],
        [18],
        [ 3],
        [ 9],
        [ 2],
        [ 7],
        [10],
        [ 6],
        [14],
        [ 4],
        [ 2],
        [ 2],
        [ 3],
        [15],
        [22],
        [18],
        [ 2],
        [ 3],
        [ 4],
        [ 7],
        [18],
        [18],
        [ 6],
        [16],
        [ 2],
        [ 2],
        [ 8],
        [ 4],
        [ 5],
        [ 4],
        [ 5],
        [ 5],
        [12],
        [ 2],
        [ 2],
        [11],
        [ 3],
        [ 3],
        [ 8],
        [ 2],
        [ 5],
        [ 3],
        [15],
        [ 2],
        [ 2],
        [11],
        [ 7]])>,
 <tf.Tensor: shape=(64, 1), dtype=int32, numpy=
 array([[ 8],
        [ 8],
        [ 9],
        [ 7],
        [16],
        [ 8],
        [15],
        [ 9],
        [19],
        [11],
        [15],
        [ 9],
        [10],
        [13],
        [15],
        [ 8],
        [11],
        [ 7],
        [ 8],
        [ 9],
        [ 7],
        [11],
        [ 7],
        [13],
        [ 7],
        [10],
        [ 7],
        [12],
        [11],
        [ 7],
        [12],
        [12],
        [ 7],
        [12],
        [10],
        [ 5],
        [17],
        [11],
        [11],
        [ 7],
        [14],
        [13],
        [ 9],
        [10],
        [17],
        [10],
        [12],
        [ 8],
        [15],
        [15],
        [ 8],
        [15],
        [15],
        [ 8],
        [12],
        [ 9],
        [ 6],
        [ 5],
        [11],
        [12],
        [12],
        [ 8],
        [ 8],
        [10]], dtype=int32)>)
```

**Comment by rnyak** [comment]
> @rnyak, Can you share the code you are using to get sparse outputs? Also, could you check the `sparse_names` arg and make sure the column names match?
> 
> I am getting this output when I use the function `get_dataloader` defined in the notebook image

Sara my script was `next(iter(get_dataloader(train_paths)))[0]` but I dont get sparse tensors now after I pulled your latest changes today. The batch is dense now.

**Comment by sararb** [comment]
Close the PR to break it into small independent PRs for easy tracking/reviewing
